doc/adr004-effect-system.md

```diff
@@ -38,150 +36,292 @@ simplicity.
+structured as a list of effect components. An effect component has a kind and an
+entity. The kind can be `read`, `update` or `temporal`, and the entity is the
+entity the effect acts upon.
+
+```typescript
+/* Kinds for concrete effect components */
+export type ComponentKind = 'read' | 'update' | 'temporal'
+
+/* A concrete effect component, specifying a kind and the entity it acts upon */
+export interface EffectComponent { kind: ComponentKind, entity: Entity }
+
+/* Concrete atomic effects specifying which kinds of effects act upon which entities */
+export interface ConcreteEffect { kind: 'concrete', components: EffectComponent[] }
+
 ```
 
-Other forms are actually sugaring for these two. Effects like `Update['x'] &
-Update['y']` are in an invalid format and should be simplified before applying
-any other transformation. See equivalence rules below.
+The effect `Read['x', 'y'] & Update[v]` would be represented as:
 
-The motivation for this form is to help writing effect signatures for operators
-that care only about the read or the update part of some effect. For example,
-the `or` operator representing disjunction in Quint takes expressions with the
-identical updates, but doesn't have any restrictions on the read part. Ensuring
-this normal form allows us to write its signature as `(Read[r1] & Update[u],
-Read[r2] & Update[u]) => Read[r1 ∪ r2] & Update[u]`.
+```typescript
+{
+  kind: 'concrete',
+  components: [
+    { kind: 'read', entity: { kind: 'concrete', stateVariables: [{ name: 'x', reference: 1n }, { name: 'y', reference: 2n }] } },
+    { kind: 'update', entity: { kind: 'variable', name: 'v' } },
+  ],
+}
+```
 
-### Equivalence rules
+Effect signatures for operators may depend on its parameters. For that purpose,
+we also define arrow effects, which carry a list of effects for the respective
+parameters of an operator and the resulting effect.
 
-These are some equivalence rules to be used alongside unification, but that
-don't require any substitution. These are applied in a simplification process
-with the goal of reaching the normal form. Equivalence between `E1` and `E2` is
-expressed by `E1 ≡ E2`, and the equivalence symbol `≡` has the lowest precedence
-on this system.
+```typescript
+/* Arrow effects for expressions with effects depending on parameters */
+export interface ArrowEffect { kind: 'arrow', params: Effect[], result: Effect }
+```
 
+Just like with entities, an effect can also be a variable, which can be
+substituted with another effect during inferrence. Those are the most general
+kinds of effects, and will be used when no effect information for a given
+expression is available, i.e. `def foo(x) = x` will have a `(e) => e` effect
+signature, where `e` is an effect variable, since there are no constraints on
+the effect of `x`. An effect variable is equivalent to `Read[v1] & Update[v2] &
+Temporal[v3]`, where `v1`, `v2` and `v3` are entity variables.
```

Comment:
I think that possibly part of the conceptual confusion i am hitting in this paragraph results from not keeping a clear separation between the ts representation forming our meta language, and the object language? So, `EffectVariable` is one of the types comprising the `Effect` union type`, but an effect variable *stands for* but *is not itself* an effect. Does this make sense?

This would suggest a rephrasing like:

```suggestion
When no effect information for a given expression is available, the effect is left
undetermined, as indicated by an effect variable. I.e. `def foo(x) = x` will have be inferred to have the effect signature `(e) => e`, where `e` is an effect variable, since there are no constraints on the effect of `x`. An **effect variable** `e` is equivalent to the effect `Read[v1] & Update[v2] & Temporal[v3]`, where `v1`, `v2` and `v3` are _entity variables_.
```